packages/grafana-ui/src/components/GraphNG/utils.ts

```diff
@@ -2,38 +2,72 @@ import { XYFieldMatchers } from './types';
 import { ArrayVector, DataFrame, FieldType, outerJoinDataFrames } from '@grafana/data';
 import { nullToUndefThreshold } from './nullToUndefThreshold';
 
-function applySpanNullsThresholds(frames: DataFrame[]) {
-  for (const frame of frames) {
-    let refField = frame.fields.find((field) => field.type === FieldType.time); // this doesnt need to be time, just any numeric/asc join field
-    let refValues = refField?.values.toArray() as any[];
+function applySpanNullsThresholds(frame: DataFrame) {
+  let refField = frame.fields.find((field) => field.type === FieldType.time); // this doesnt need to be time, just any numeric/asc join field
+  let refValues = refField?.values.toArray() as any[];
 
-    for (let i = 0; i < frame.fields.length; i++) {
-      let field = frame.fields[i];
+  for (let i = 0; i < frame.fields.length; i++) {
+    let field = frame.fields[i];
 
-      if (field === refField) {
-        continue;
+    if (field === refField) {
+      continue;
+    }
+
+    if (field.type === FieldType.number) {
+      let spanNulls = field.config.custom?.spanNulls;
+
+      if (typeof spanNulls === 'number') {
+        field.values = new ArrayVector(nullToUndefThreshold(refValues, field.values.toArray(), spanNulls));
       }
+    }
+  }
+}
+
+// sets to undefined all same future values not preceeded by explicit null
+// in:  1,        1,undefined,        1,2,        2,null,2,3
+// out: 1,undefined,undefined,undefined,2,undefined,null,2,3
+function unsetSameFutureValues(frame: DataFrame) {
```

Comment:
Would be nice with unit tests for this function

Reply:
agreed. i'm considering removing the join/alignment step for the timeline panel since it uses quadtree for cursor hover detection as well as a custom renderer. that will affect how extensive the tests for this function need to be, since it may not need to handle `undefined` values that get inserted as part of the join/alignment step.

i will add a couple tests before merging, though.

Reply:
I copied this into:
https://github.com/grafana/grafana/pull/34171

and call it from the main panel component